# Implementing Various Gradient Descent Without Feature Scaling and with Feature Scaling to know the importance of feature scaling in Gradient Descent

In [2]:
import numpy as np
import pandas as pd
a=1
b=np.array([9,8,7,6,5])
a-b

array([-8, -7, -6, -5, -4])

In [4]:
data=np.loadtxt("turbo.csv",delimiter=",")
data.shape

(7176, 5)

In [5]:
df=pd.DataFrame(data)
df.columns=["Temperature","Ambient_pressure","Relative_humidity","Exhaust_vaccume","Electrical_output"]
df.isnull().sum()

Temperature          0
Ambient_pressure     0
Relative_humidity    0
Exhaust_vaccume      0
Electrical_output    0
dtype: int64

In [6]:
df.head()

Temperature  Ambient_pressure  Relative_humidity  Exhaust_vaccume  \
0         8.58             38.38            1021.03            84.37   
1        21.79             58.20            1017.21            66.74   
2        16.64             48.92            1011.55            78.76   
3        31.38             71.32            1009.17            60.42   
4         9.20             40.03            1017.05            92.46   

   Electrical_output  
0             482.26  
1             446.94  
2             452.56  
3             433.44  
4             480.38

In [7]:
df.describe()
df.iloc[0][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']]

Temperature             8.58
Ambient_pressure       38.38
Relative_humidity    1021.03
Exhaust_vaccume        84.37
Name: 0, dtype: float64

###  Batch Gradient Descent Without Feature Scaling......... with a learning rate=0.0000001 and 1500 iterations to converge the function for optimal coefficents....

In [8]:
def Gradient_Descent(data):
    m=[0,0,0,0]
    c=0
    learning_rate=0.0001
    no_of_iterations=3000
    for i in range(no_of_iterations):
        m,c=Step_Gradient(data,m,c,learning_rate)
        print("The  m is {} and c is {} and cost {}".format(m,c,cost(data,m,c)))
    return m,c

In [9]:
def cost(data,m,c):
    M=len(data)
    cost=0
    for i in range(M):
        
        x=np.array(data.iloc[i][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']])
        y=np.array(data.iloc[i][['Electrical_output']])
        cost=cost+(1/M)*(y-np.sum(m*x)-c)**2
    return cost
    

In [10]:
def Step_Gradient(data,m,c,learning_rate):
    m_slope=[0,0,0,0]
    c_slope=0
    M=len(data)
    for i in range(len(data)):
        x=np.array(data.iloc[i][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']])
        y=np.array(data.iloc[i][['Electrical_output']])
        m_slope=m_slope+(-2/M)*(y-np.sum(m*x)-c)*x
        c_slope=c_slope+(-2/M)*(y-np.sum(m*x)-c)
    m=m-learning_rate*m_slope
    c=c-learning_rate*c_slope
    return m,c       

## Clearly from the results we can observe the cost function is decreasing which means i escaped from the problem of Overshooting so my parameters of learning rate and no of iterations seems to be appropriate here.But converging to optima may take more number of iterations.Learning reat choosen is 0.0000001 and no of iterations=3000.

In [ ]:
m,c=Gradient_Descent(df)
print("Final m {} and c {} are".format(m,c))

The  m is [ 0.015838    0.04406446  0.8289208   0.06011513] and c is [ 0.00081798] and cost [ 154366.21217994]
The  m is [ 0.00178004  0.00539743  0.11292672  0.00848085] and c is [ 0.00011131] and cost [ 115242.6645643]
The  m is [ 0.01354603  0.03819321  0.731413    0.05337425] and c is [ 0.00072161] and cost [ 86051.31320014]
The  m is [ 0.00300574  0.00926047  0.19718992  0.0148879 ] and c is [ 0.00019433] and cost [ 64270.68436114]
The  m is [ 0.01173353  0.03364841  0.65866441  0.04842384] and c is [ 0.00064968] and cost [ 48019.42758342]
The  m is [ 0.00381792  0.01197881  0.26006638  0.01974752] and c is [ 0.00025625] and cost [ 35893.80840366]
The  m is [ 0.01027886  0.03009346  0.60438946  0.04480925] and c is [ 0.00059599] and cost [ 26846.45660491]
The  m is [ 0.00432165  0.01384318  0.30698551  0.02345242] and c is [ 0.00030242] and cost [ 20095.89648784]
The  m is [ 0.00909126  0.02727722  0.56389825  0.04219122] and c is [ 0.00055589] and cost [ 15059.04457465]
The  m is

The  m is [ -6.54137840e-03  -4.42072166e-04   4.45490881e-01   4.37796545e-02] and c is [ 0.0004346] and cost [ 243.83755105]
The  m is [-0.00673725 -0.00075523  0.44551292  0.04392902] and c is [ 0.00043456] and cost [ 243.6608296]
The  m is [-0.00693348 -0.00106946  0.44551226  0.04407663] and c is [ 0.00043449] and cost [ 243.48427918]
The  m is [-0.00712925 -0.0013825   0.4455312   0.04422554] and c is [ 0.00043445] and cost [ 243.30789888]
The  m is [-0.00732527 -0.00169632  0.44553321  0.04437313] and c is [ 0.00043438] and cost [ 243.131688]
The  m is [-0.00752092 -0.00200921  0.44554984  0.04452166] and c is [ 0.00043433] and cost [ 242.95564597]
The  m is [-0.00771673 -0.00232264  0.44555384  0.04466917] and c is [ 0.00043427] and cost [ 242.7797723]
The  m is [-0.00791225 -0.00263535  0.44556875  0.04481736] and c is [ 0.00043422] and cost [ 242.60406662]
The  m is [-0.00810787 -0.00294842  0.44557424  0.04496476] and c is [ 0.00043416] and cost [ 242.42852858]
The  m is [-0

The  m is [-0.02121434 -0.02390207  0.4462437   0.05475682] and c is [ 0.00043047] and cost [ 230.87518405]
The  m is [-0.02140432 -0.02420557  0.44625348  0.05489716] and c is [ 0.00043041] and cost [ 230.7107868]
The  m is [-0.02159422 -0.02450893  0.44626326  0.05503739] and c is [ 0.00043036] and cost [ 230.54654542]
The  m is [-0.02178405 -0.02481216  0.44627303  0.05517751] and c is [ 0.00043031] and cost [ 230.38245975]
The  m is [-0.02197379 -0.02511526  0.44628281  0.05531753] and c is [ 0.00043025] and cost [ 230.21852964]
The  m is [-0.02216346 -0.02541822  0.44629258  0.05545745] and c is [ 0.0004302] and cost [ 230.05475494]
The  m is [-0.02235305 -0.02572105  0.44630235  0.05559726] and c is [ 0.00043015] and cost [ 229.89113548]
The  m is [-0.02254255 -0.02602375  0.44631212  0.05573696] and c is [ 0.00043009] and cost [ 229.72767111]
The  m is [-0.02273198 -0.02632632  0.44632189  0.05587657] and c is [ 0.00043004] and cost [ 229.56436168]
The  m is [-0.02292133 -0.0266

The  m is [-0.03543027 -0.04659243  0.44698371  0.06512794] and c is [ 0.00042648] and cost [ 218.81449499]
The  m is [-0.03561435 -0.04688598  0.4469934   0.06526048] and c is [ 0.00042643] and cost [ 218.66151185]
The  m is [-0.03579835 -0.0471794   0.4470031   0.06539293] and c is [ 0.00042637] and cost [ 218.50867322]
The  m is [-0.03598228 -0.04747269  0.44701279  0.06552527] and c is [ 0.00042632] and cost [ 218.35597894]
The  m is [-0.03616613 -0.04776585  0.44702248  0.06565752] and c is [ 0.00042627] and cost [ 218.20342887]
The  m is [-0.03634991 -0.04805888  0.44703217  0.06578966] and c is [ 0.00042622] and cost [ 218.05102288]
The  m is [-0.03653361 -0.04835178  0.44704186  0.06592171] and c is [ 0.00042617] and cost [ 217.89876081]
The  m is [-0.03671723 -0.04864456  0.44705155  0.06605366] and c is [ 0.00042612] and cost [ 217.74664252]
The  m is [-0.03690078 -0.04893721  0.44706123  0.0661855 ] and c is [ 0.00042607] and cost [ 217.59466788]
The  m is [-0.03708425 -0.04

The  m is [-0.04920577 -0.06853968  0.44771728  0.07492014] and c is [ 0.00042264] and cost [ 207.58968291]
The  m is [-0.04938417 -0.06882362  0.44772689  0.07504525] and c is [ 0.00042259] and cost [ 207.44728223]
The  m is [-0.04956249 -0.06910744  0.44773649  0.07517025] and c is [ 0.00042254] and cost [ 207.30501553]
The  m is [-0.04974073 -0.06939114  0.4477461   0.07529516] and c is [ 0.00042249] and cost [ 207.16288269]
The  m is [-0.04991891 -0.06967471  0.4477557   0.07541998] and c is [ 0.00042244] and cost [ 207.02088356]
The  m is [-0.05009701 -0.06995815  0.44776531  0.0755447 ] and c is [ 0.00042239] and cost [ 206.87901801]
The  m is [-0.05027503 -0.07024148  0.44777491  0.07566932] and c is [ 0.00042234] and cost [ 206.73728592]
The  m is [-0.05045299 -0.07052468  0.44778451  0.07579385] and c is [ 0.00042229] and cost [ 206.59568714]
The  m is [-0.05063087 -0.07080775  0.44779411  0.07591828] and c is [ 0.00042224] and cost [ 206.45422155]
The  m is [-0.05080867 -0.07

The  m is [-0.06255694 -0.08976969  0.44844425  0.08415904] and c is [ 0.00041894] and cost [ 197.13990012]
The  m is [-0.06272986 -0.09004436  0.44845377  0.08427703] and c is [ 0.00041889] and cost [ 197.0073123]
The  m is [-0.0629027  -0.09031892  0.44846329  0.08439492] and c is [ 0.00041884] and cost [ 196.87484873]
The  m is [-0.06307548 -0.09059335  0.44847281  0.08451273] and c is [ 0.0004188] and cost [ 196.74250929]
The  m is [-0.06324818 -0.09086767  0.44848232  0.08463044] and c is [ 0.00041875] and cost [ 196.61029385]
The  m is [-0.06342081 -0.09114186  0.44849184  0.08474806] and c is [ 0.0004187] and cost [ 196.47820229]
The  m is [-0.06359337 -0.09141594  0.44850135  0.08486559] and c is [ 0.00041865] and cost [ 196.34623448]
The  m is [-0.06376586 -0.09168989  0.44851087  0.08498303] and c is [ 0.0004186] and cost [ 196.21439031]
The  m is [-0.06393828 -0.09196373  0.44852038  0.08510038] and c is [ 0.00041856] and cost [ 196.08266965]
The  m is [-0.06411063 -0.092237

The  m is [-0.07566691 -0.11057312  0.44917391  0.09298035] and c is [ 0.00041533] and cost [ 187.28534977]
The  m is [-0.07583448 -0.11083873  0.44918334  0.09309145] and c is [ 0.00041529] and cost [ 187.16197783]
The  m is [-0.07600198 -0.11110423  0.44919277  0.09320246] and c is [ 0.00041524] and cost [ 187.03872101]
The  m is [-0.07616941 -0.11136961  0.4492022   0.09331339] and c is [ 0.00041519] and cost [ 186.9155792]
The  m is [-0.07633677 -0.11163487  0.44921162  0.09342423] and c is [ 0.00041515] and cost [ 186.79255229]
The  m is [-0.07650407 -0.11190002  0.44922105  0.09353498] and c is [ 0.0004151] and cost [ 186.66964016]
The  m is [-0.07667129 -0.11216506  0.44923047  0.09364564] and c is [ 0.00041506] and cost [ 186.5468427]
The  m is [-0.07683845 -0.11242998  0.4492399   0.09375622] and c is [ 0.00041501] and cost [ 186.42415981]
The  m is [-0.07700555 -0.11269478  0.44924932  0.09386671] and c is [ 0.00041496] and cost [ 186.30159135]
The  m is [-0.07717257 -0.11295

The  m is [-0.08837269 -0.13069088  0.44989653  0.10128333] and c is [ 0.00041186] and cost [ 178.11439463]
The  m is [-0.08853511 -0.13094775  0.44990587  0.10138785] and c is [ 0.00041182] and cost [ 177.99956258]
The  m is [-0.08869746 -0.13120451  0.4499152   0.1014923 ] and c is [ 0.00041177] and cost [ 177.88483722]
The  m is [-0.08885975 -0.13146115  0.44992454  0.10159666] and c is [ 0.00041173] and cost [ 177.77021845]
The  m is [-0.08902197 -0.13171768  0.44993387  0.10170093] and c is [ 0.00041168] and cost [ 177.65570615]
The  m is [-0.08918413 -0.1319741   0.4499432   0.10180512] and c is [ 0.00041164] and cost [ 177.54130023]
The  m is [-0.08934622 -0.13223041  0.44995254  0.10190923] and c is [ 0.0004116] and cost [ 177.42700059]
The  m is [-0.08950824 -0.13248661  0.44996187  0.10201326] and c is [ 0.00041155] and cost [ 177.3128071]
The  m is [-0.0896702 -0.1327427  0.4499712  0.1021172] and c is [ 0.00041151] and cost [ 177.19871968]
The  m is [-0.0898321  -0.13299867

The  m is [-0.1006892  -0.15014702  0.45061198  0.10909142] and c is [ 0.00040852] and cost [ 169.57694567]
The  m is [-0.10084666 -0.15039544  0.45062122  0.10918967] and c is [ 0.00040848] and cost [ 169.47002847]
The  m is [-0.10100405 -0.15064376  0.45063046  0.10928784] and c is [ 0.00040844] and cost [ 169.36321015]
The  m is [-0.10116138 -0.15089198  0.45063971  0.10938593] and c is [ 0.00040839] and cost [ 169.25649063]
The  m is [-0.10131865 -0.15114008  0.45064895  0.10948394] and c is [ 0.00040835] and cost [ 169.1498698]
The  m is [-0.10147585 -0.15138808  0.45065818  0.10958187] and c is [ 0.00040831] and cost [ 169.04334757]
The  m is [-0.101633   -0.15163597  0.45066742  0.10967972] and c is [ 0.00040827] and cost [ 168.93692385]
The  m is [-0.10179008 -0.15188375  0.45067666  0.10977749] and c is [ 0.00040822] and cost [ 168.83059853]
The  m is [-0.10194709 -0.15213143  0.45068589  0.10987519] and c is [ 0.00040818] and cost [ 168.72437151]
The  m is [-0.10210405 -0.152

The  m is [-0.11263075 -0.16896467  0.45132015  0.11642709] and c is [ 0.00040531] and cost [ 161.6266851]
The  m is [-0.11278343 -0.16920495  0.4513293   0.11651935] and c is [ 0.00040527] and cost [ 161.52710472]
The  m is [-0.11293605 -0.16944513  0.45133845  0.11661153] and c is [ 0.00040522] and cost [ 161.42761602]
The  m is [-0.11308861 -0.16968521  0.45134759  0.11670364] and c is [ 0.00040518] and cost [ 161.3282189]
The  m is [-0.11324111 -0.16992518  0.45135674  0.11679567] and c is [ 0.00040514] and cost [ 161.22891328]
The  m is [-0.11339355 -0.17016505  0.45136588  0.11688762] and c is [ 0.0004051] and cost [ 161.12969907]
The  m is [-0.11354592 -0.17040482  0.45137503  0.1169795 ] and c is [ 0.00040506] and cost [ 161.03057617]
The  m is [-0.11369824 -0.17064448  0.45138417  0.1170713 ] and c is [ 0.00040502] and cost [ 160.9315445]
The  m is [-0.11385049 -0.17088404  0.45139331  0.11716302] and c is [ 0.00040498] and cost [ 160.83260397]
The  m is [-0.11400269 -0.171123

The  m is [-0.12406301 -0.18693361  0.4520119   0.12322525] and c is [ 0.00040225] and cost [ 154.31364172]
The  m is [-0.12421114 -0.18716614  0.45202096  0.12331186] and c is [ 0.00040221] and cost [ 154.22077851]
The  m is [-0.12435922 -0.18739856  0.45203001  0.1233984 ] and c is [ 0.00040217] and cost [ 154.12800038]
The  m is [-0.12450724 -0.18763089  0.45203906  0.12348486] and c is [ 0.00040213] and cost [ 154.03530725]
The  m is [-0.1246552  -0.18786311  0.45204811  0.12357126] and c is [ 0.0004021] and cost [ 153.94269905]
The  m is [-0.12480311 -0.18809524  0.45205716  0.12365758] and c is [ 0.00040206] and cost [ 153.85017569]
The  m is [-0.12495095 -0.18832726  0.45206621  0.12374383] and c is [ 0.00040202] and cost [ 153.75773708]
The  m is [-0.12509873 -0.18855919  0.45207525  0.12383   ] and c is [ 0.00040198] and cost [ 153.66538315]
The  m is [-0.12524646 -0.18879101  0.4520843   0.12391611] and c is [ 0.00040194] and cost [ 153.5731138]
The  m is [-0.12539413 -0.1890

The  m is [-0.13529993 -0.20454796  0.45270535  0.12968518] and c is [ 0.00039928] and cost [ 147.40615837]
The  m is [-0.13544363 -0.20477289  0.4527143   0.12976634] and c is [ 0.00039924] and cost [ 147.31960914]
The  m is [-0.13558727 -0.20499773  0.45272326  0.12984742] and c is [ 0.0003992] and cost [ 147.23313882]
The  m is [-0.13573086 -0.20522247  0.45273221  0.12992843] and c is [ 0.00039916] and cost [ 147.14674734]
The  m is [-0.13587439 -0.20544712  0.45274117  0.13000937] and c is [ 0.00039912] and cost [ 147.06043461]
The  m is [-0.13601787 -0.20567167  0.45275012  0.13009025] and c is [ 0.00039909] and cost [ 146.97420056]
The  m is [-0.13616129 -0.20589612  0.45275907  0.13017105] and c is [ 0.00039905] and cost [ 146.88804512]
The  m is [-0.13630465 -0.20612048  0.45276802  0.13025179] and c is [ 0.00039901] and cost [ 146.8019682]
The  m is [-0.13644796 -0.20634474  0.45277696  0.13033246] and c is [ 0.00039897] and cost [ 146.71596973]
The  m is [-0.13659121 -0.2065

The  m is [-0.14745443 -0.22354313  0.45347094  0.13641545] and c is [ 0.00039609] and cost [ 140.24361713]
The  m is [-0.14759337 -0.2237599   0.45347979  0.1364908 ] and c is [ 0.00039605] and cost [ 140.16358026]
The  m is [-0.14773225 -0.22397657  0.45348863  0.13656608] and c is [ 0.00039601] and cost [ 140.08361595]
The  m is [-0.14787108 -0.22419316  0.45349748  0.1366413 ] and c is [ 0.00039598] and cost [ 140.00372413]
The  m is [-0.14800986 -0.22440965  0.45350632  0.13671645] and c is [ 0.00039594] and cost [ 139.92390473]
The  m is [-0.14814858 -0.22462605  0.45351516  0.13679154] and c is [ 0.0003959] and cost [ 139.84415767]
The  m is [-0.14828725 -0.22484236  0.453524    0.13686656] and c is [ 0.00039587] and cost [ 139.76448288]
The  m is [-0.14842586 -0.22505857  0.45353284  0.13694152] and c is [ 0.00039583] and cost [ 139.6848803]
The  m is [-0.14856442 -0.22527469  0.45354168  0.13701641] and c is [ 0.0003958] and cost [ 139.60534985]
The  m is [-0.14870293 -0.22549

The  m is [-0.15799601 -0.23996525  0.45414837  0.14202814] and c is [ 0.00039334] and cost [ 134.28800594]
The  m is [-0.15813085 -0.24017498  0.45415712  0.14209855] and c is [ 0.00039331] and cost [ 134.2133543]
The  m is [-0.15826563 -0.24038462  0.45416586  0.1421689 ] and c is [ 0.00039327] and cost [ 134.13876997]
The  m is [-0.15840037 -0.24059416  0.45417461  0.14223918] and c is [ 0.00039324] and cost [ 134.06425289]
The  m is [-0.15853505 -0.24080362  0.45418335  0.1423094 ] and c is [ 0.0003932] and cost [ 133.98980301]
The  m is [-0.15866969 -0.24101299  0.45419209  0.14237956] and c is [ 0.00039317] and cost [ 133.91542024]
The  m is [-0.15880426 -0.24122227  0.45420084  0.14244966] and c is [ 0.00039313] and cost [ 133.84110453]
The  m is [-0.15893879 -0.24143146  0.45420958  0.14251969] and c is [ 0.0003931] and cost [ 133.76685582]
The  m is [-0.15907327 -0.24164056  0.45421831  0.14258966] and c is [ 0.00039306] and cost [ 133.69267403]
The  m is [-0.15920769 -0.24184

The  m is [-0.16822769 -0.25585438  0.45481814  0.14726919] and c is [ 0.00039071] and cost [ 128.73210662]
The  m is [-0.16835858 -0.2560573   0.45482679  0.14733488] and c is [ 0.00039067] and cost [ 128.66245186]
The  m is [-0.16848943 -0.25626015  0.45483544  0.14740052] and c is [ 0.00039064] and cost [ 128.59285958]
The  m is [-0.16862022 -0.2564629   0.45484408  0.1474661 ] and c is [ 0.00039061] and cost [ 128.5233297]
The  m is [-0.16875096 -0.25666557  0.45485273  0.14753162] and c is [ 0.00039057] and cost [ 128.45386217]
The  m is [-0.16888165 -0.25686815  0.45486137  0.14759708] and c is [ 0.00039054] and cost [ 128.38445694]
The  m is [-0.16901229 -0.25707065  0.45487001  0.14766248] and c is [ 0.0003905] and cost [ 128.31511393]
The  m is [-0.16914287 -0.25727305  0.45487865  0.14772782] and c is [ 0.00039047] and cost [ 128.24583309]
The  m is [-0.16927341 -0.25747538  0.45488729  0.1477931 ] and c is [ 0.00039044] and cost [ 128.17661436]
The  m is [-0.1694039  -0.2576

### Applying directly the linear Regression algorithm from linear model package of sklearn to see the optimal coefficents and score of training data.The score here in linear regression is my coefficent of determination...

In [20]:
from sklearn.linear_model import LinearRegression
alg1=LinearRegression()
alg1.fit(df[['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']],df[['Electrical_output']])
m=alg1.coef_
c=alg1.intercept_

In [30]:
alg1.score(df[['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']],df[['Electrical_output']])

0.92876320004405988

In [22]:
test_data=np.loadtxt("turbaintest.csv",delimiter=",")

ysklearnpred=[]
yGraddescpred=[]
for i in range(len(test_data)):
    ysklearnpred.extend(np.dot(np.array(alg1.coef_),test_data[i])+alg1.intercept_)
    yGraddescpred.extend(np.dot(np.array(m),test_data[i])+c)
np.savetxt('Turbainsklelarn.csv',ysklearnpred,fmt='%.5f')#-------------->Accuracy for this model is 92.8%(Based on coefficent of determination)
np.savetxt('TurbainGradient.csv',yGraddescpred,fmt='%.5f')#---------->Accuracy for this model is 23.5%




### Now i wanted to go for feature Scaling because Gradient descent is sensitive to lower values....

In [11]:
from sklearn import preprocessing
columnNames=df.columns
print(columnNames)
data = pd.DataFrame(preprocessing.scale(df),columns=df.columns)
data.head()

Index(['Temperature', 'Ambient_pressure', 'Relative_humidity',
       'Exhaust_vaccume', 'Electrical_output'],
      dtype='object')


Temperature  Ambient_pressure  Relative_humidity  Exhaust_vaccume  \
0    -1.478275         -1.247642           1.302211         0.758625   
1     0.289012          0.306798           0.661749        -0.446922   
2    -0.399976         -0.421013          -0.287207         0.375011   
3     1.572001          1.335770          -0.686238        -0.879086   
4    -1.395329         -1.118236           0.634923         1.311822   

   Electrical_output  
0           1.624239  
1          -0.437234  
2          -0.109219  
3          -1.225170  
4           1.514512

### I wanted to apply sklearn linearRegression to see the optimal coefficients from sklearn after column standardisation...

In [23]:
from sklearn.linear_model import LinearRegression
alg2=LinearRegression()
alg2.fit(data[['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']],data[['Electrical_output']])
print(alg2.coef_)
print(alg2.intercept_)

[[-0.86980688 -0.16892631  0.02041279 -0.13670987]]
[  3.21780987e-14]


### I applied Gradient Descent on the feature scaled data to see how my algorithm converge to optimal values.To my surprise the cost function decreased drastically and converge to optima with very less no of iterations equal to 20.After that from the results down below the cost remains same.The optimal values were changing.I had choosen the learning_rate as 0.01. and no of iterations =100 but my algorithm converge to the global optima within less no of iterations.And my coefficient i achieve through sklearn and gradient descent are almost equal.

In [23]:
feature_scaler_m,feature_scaler_c=Gradient_Descent(data)
feature_scaler_m,feature_scaler_c

The  m is [-0.56907426 -0.52177186  0.31193242  0.23579269] and c is [  2.11409097e-14] and cost [ 0.30989843]
The  m is [-0.35961192 -0.31907026  0.11787442  0.02614296] and c is [  5.97401604e-14] and cost [ 0.15019985]
The  m is [-0.5065166  -0.43252728  0.16774434  0.06100209] and c is [  5.42676825e-14] and cost [ 0.10996689]
The  m is [-0.48109057 -0.38430711  0.11228688  0.00236341] and c is [  5.64219651e-14] and cost [ 0.09764452]
The  m is [ -5.31505203e-01  -4.02906478e-01   1.13511849e-01  -2.69225780e-04] and c is [  5.11703920e-14] and cost [ 0.09237816]
The  m is [-0.54272409 -0.38496621  0.09410443 -0.02141373] and c is [  4.88914288e-14] and cost [ 0.08920397]
The  m is [-0.56915805 -0.38096318  0.08869282 -0.02896473] and c is [  4.58215334e-14] and cost [ 0.08683796]
The  m is [-0.58588576 -0.36873883  0.07991506 -0.03954685] and c is [  4.39333872e-14] and cost [ 0.08489705]
The  m is [-0.60492106 -0.35956612  0.07499883 -0.04640416] and c is [  4.22309323e-14] and 

The  m is [-0.86568042 -0.1719804   0.02118079 -0.13536446] and c is [  3.22876694e-14] and cost [ 0.07123973]
The  m is [-0.86593237 -0.17179393  0.0211339  -0.1354466 ] and c is [  3.22790419e-14] and cost [ 0.07123938]
The  m is [-0.86616894 -0.17161884  0.02108987 -0.13552373] and c is [  3.22778140e-14] and cost [ 0.07123907]
The  m is [-0.86639105 -0.17145444  0.02104853 -0.13559616] and c is [  3.22691540e-14] and cost [ 0.07123881]
The  m is [-0.86659961 -0.17130008  0.02100971 -0.13566416] and c is [  3.22616323e-14] and cost [ 0.07123857]
The  m is [-0.86679544 -0.17115515  0.02097326 -0.135728  ] and c is [  3.22555825e-14] and cost [ 0.07123836]
The  m is [-0.86697931 -0.17101907  0.02093904 -0.13578795] and c is [  3.22509882e-14] and cost [ 0.07123817]
The  m is [-0.86715195 -0.17089129  0.02090691 -0.13584424] and c is [  3.22469305e-14] and cost [ 0.07123801]
The  m is [-0.86731405 -0.17077131  0.02087674 -0.13589709] and c is [  3.22348634e-14] and cost [ 0.07123787]
T

The  m is [-0.86977027 -0.1689534   0.0204196  -0.13669793] and c is [  3.21713237e-14] and cost [ 0.0712368]
The  m is [-0.86977251 -0.16895175  0.02041918 -0.13669866] and c is [  3.21638834e-14] and cost [ 0.0712368]
The  m is [-0.86977461 -0.16895019  0.02041879 -0.13669934] and c is [  3.21641517e-14] and cost [ 0.0712368]
The  m is [-0.86977658 -0.16894874  0.02041843 -0.13669999] and c is [  3.21672742e-14] and cost [ 0.0712368]
The  m is [-0.86977843 -0.16894737  0.02041808 -0.13670059] and c is [  3.21715921e-14] and cost [ 0.0712368]
The  m is [-0.86978016 -0.16894608  0.02041776 -0.13670116] and c is [  3.21639484e-14] and cost [ 0.0712368]
The  m is [-0.86978179 -0.16894487  0.02041745 -0.13670169] and c is [  3.21679898e-14] and cost [ 0.0712368]
The  m is [-0.86978333 -0.16894374  0.02041717 -0.13670219] and c is [  3.21724703e-14] and cost [ 0.0712368]
The  m is [-0.86978476 -0.16894268  0.0204169  -0.13670266] and c is [  3.21674125e-14] and cost [ 0.0712368]
The  m is 

KeyboardInterrupt: 

In [24]:
test_data

array([[   11.95,    42.03,  1017.58,    90.89],
       [   12.07,    38.25,  1012.67,    81.66],
       [   26.91,    74.99,  1005.64,    78.98],
       ..., 
       [   24.32,    66.25,  1009.09,    91.89],
       [   23.49,    42.8 ,  1013.96,    65.31],
       [   21.76,    60.27,  1018.96,    85.06]])

In [27]:
result=alg2.predict(test_data)

### when i tried to build a model intially using my own Gradient descent  score was 23.3% which is pretty bad.But i applied feature scaling and bring up my score to 92.8% which is great...

In [29]:
alg2.score(data[['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']],data[['Electrical_output']])

0.92876320004405988

### I wanted to apply stochastic Gradient descent but to my surprise even after choosing a learning_rate of 0.002 and i consistently changed my m and values point by point since i have 7176 data points i tried to change my m and c values converge to optima fastly and never stay there as it happened in Batch Gradient descent.I choose to run gradient descent for only one iteration because the dataset is larger.Frequently updataing m and c will lead to optimal fastly.

In [12]:
def stochastic_gd(data):
    no_of_iterations=100
    learning_rate=0.1
    m=np.zeros(4)
    c=0
    m,c=step_stochastic_gradient(data,learning_rate,m,c)
    return m,c

In [13]:
def step_stochastic_gradient(data,learning_rate,m,c):
    M=len(data)
    m_slope=np.zeros(4)
    c_slope=0
    count=0
    prevcost=1000
    for i in range(len(data)):
        x=np.array(data.iloc[i][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']])
        y=np.array(data.iloc[i]['Electrical_output'])
        m_slope=m_slope+((-2/M)*(y-np.sum(m*x))*x)
        c_slope=c_slope+((-2/M)*(y-np.sum(m*x)))
        m=m-(learning_rate*m_slope)
        c=c-(learning_rate*c_slope)
        costvalue=stochastic_cost(data,m,c)
       
        print("The m is {} and c is {} and cost is {}".format(m,c,costvalue))
         
    return m,c

In [14]:
def stochastic_cost(data,m,c):
    cost=0
    for i in range(len(data)):
        x=np.array(data.iloc[i][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']])
        y=np.array(data.iloc[i]['Electrical_output'])
        cost=cost+(-2/len(data))*(y-np.sum(m*x)-c)**2
    return cost

In [ ]:
m,c=stochastic_gd(data)
print("Arrived to final m values are: {} and c is {}".format(m,c))

The m is [ -1.87374625e-07  -1.58141374e-07   1.65058121e-07   9.61573665e-08] and c is 1.2675224044194023e-07 and cost is -1.9999982446406348
The m is [ -3.84610575e-07  -3.26750930e-07   3.07536825e-07   2.07564070e-07] and c is 2.1938366727333193e-07 and cost is -1.9999964384405597
The m is [ -5.78437432e-07  -4.91772088e-07   4.52463471e-07   3.15774462e-07] and c is 3.034918372633231e-07 and cost is -1.99999465759096
The m is [ -9.22562484e-07  -7.84505589e-07   6.63001014e-07   5.08033824e-07] and c is 2.9199050647991134e-07 and cost is -1.9999915937328823
The m is [ -1.43160016e-06  -1.20940236e-06   9.48579582e-07   8.55336243e-07] and c is 3.986782703203973e-07 and cost is -1.9999870447342474
The m is [ -2.01981513e-06  -1.73075409e-06   1.23392400e-06   1.32880293e-06] and c is 4.2305649648537516e-07 and cost is -1.9999816620076745
The m is [ -2.79931492e-06  -2.40112604e-06   1.76860819e-06   1.85721708e-06] and c is 5.883060577120427e-07 and cost is -1.99997443034917
The m 

The m is [ -1.50362801e-04  -1.34369830e-04   9.46034572e-05   7.00843792e-05] and c is 2.084596089327231e-05 and cost is -1.9986555108914152
The m is [ -1.55412223e-04  -1.38659888e-04   9.75540398e-05   7.17823123e-05] and c is 2.138510629962966e-05 and cost is -1.9986126435091534
The m is [ -1.60506488e-04  -1.42964868e-04   1.00502358e-04   7.35697554e-05] and c is 2.1877175177331534e-05 and cost is -1.998569418782201
The m is [ -1.65668019e-04  -1.47323518e-04   1.03434720e-04   7.54730652e-05] and c is 2.2307325839530825e-05 and cost is -1.9985256040201593
The m is [ -1.71059022e-04  -1.51794500e-04   1.06516223e-04   7.73717181e-05] and c is 2.2872113607537034e-05 and cost is -1.9984802263651682
The m is [ -1.76534783e-04  -1.56357323e-04   1.09663028e-04   7.93393166e-05] and c is 2.3377190661706563e-05 and cost is -1.9984339645058593
The m is [ -1.82176512e-04  -1.61099939e-04   1.12896965e-04   8.13491639e-05] and c is 2.3782385993347623e-05 and cost is -1.9983862012905163
Th

The m is [-0.00068232 -0.00058943  0.0004106   0.0002483 ] and c is 9.270948569789115e-05 and cost is -1.9941219297883206
The m is [-0.00069294 -0.0005988   0.00041723  0.00025162] and c is 9.410975033962223e-05 and cost is -1.9940301716288371
The m is [-0.00070357 -0.00060817  0.00042386  0.00025494] and c is 9.550439032076006e-05 and cost is -1.9939384214967264
The m is [-0.00071422 -0.00061759  0.00043055  0.00025826] and c is 9.694152311392707e-05 and cost is -1.9938462848069862
The m is [-0.00072501 -0.00062715  0.00043737  0.00026163] and c is 9.848595611149644e-05 and cost is -1.99375278856999
The m is [-0.00073595 -0.00063683  0.0004443   0.00026501] and c is 9.992027724756466e-05 and cost is -1.993658079497019
The m is [-0.000747   -0.00064664  0.00045132  0.00026831] and c is 0.00010124687392594601 and cost is -1.9935624097288553
The m is [-0.00075808 -0.00065648  0.00045834  0.00027157] and c is 0.00010252723561070639 and cost is -1.9934666025954642
The m is [-0.00076923 -0.

The m is [-0.00157459 -0.00136119  0.00092919  0.00054655] and c is 0.000189336869330113 and cost is -1.986526392454139
The m is [-0.00159045 -0.00137496  0.000938    0.00055204] and c is 0.00019063700366823562 and cost is -1.9863918634755044
The m is [-0.00160633 -0.00138879  0.00094679  0.00055758] and c is 0.00019197570617063457 and cost is -1.9862570913917381
The m is [-0.00162225 -0.00140264  0.0009556   0.00056315] and c is 0.0001933566946747875 and cost is -1.9861219997192634
The m is [-0.00163826 -0.00141654  0.0009644   0.00056881] and c is 0.00019467154742096733 and cost is -1.985986239802049
The m is [-0.00165428 -0.00143045  0.00097319  0.00057446] and c is 0.00019597213443464247 and cost is -1.9858504679340747
The m is [-0.00167057 -0.00144455  0.00098193  0.00058032] and c is 0.00019743447980364424 and cost is -1.9857127841518785
The m is [-0.0016869  -0.00145868  0.00099063  0.00058623] and c is 0.00019883990167613504 and cost is -1.9855749034890493
The m is [-0.00170326

The m is [-0.00281346 -0.00244648  0.00160841  0.00103104] and c is 0.0002832759336012797 and cost is -1.9759351641458498
The m is [-0.00283416 -0.0024646   0.00161998  0.00103972] and c is 0.00028466745964839883 and cost is -1.9757571553356874
The m is [-0.00285503 -0.00248285  0.00163152  0.00104852] and c is 0.00028617137383560147 and cost is -1.975577943349856
The m is [-0.00287593 -0.00250114  0.00164311  0.0010573 ] and c is 0.0002877191099298516 and cost is -1.9753984119652146
The m is [-0.00289686 -0.00251946  0.00165466  0.00106611] and c is 0.00028921157950299604 and cost is -1.975218762422742
The m is [-0.00291784 -0.00253787  0.00166621  0.00107496] and c is 0.00029078103965615325 and cost is -1.9750384841750186
The m is [-0.00293888 -0.00255633  0.0016777   0.00108379] and c is 0.00029241771172787786 and cost is -1.9748579879109383
The m is [-0.00295995 -0.00257477  0.00168923  0.00109262] and c is 0.0002940007224672966 and cost is -1.9746773913865725
The m is [-0.00298121

The m is [-0.00439687 -0.00382944  0.0024631   0.00170267] and c is 0.00040472844899464705 and cost is -1.9624070003603022
The m is [-0.00442289 -0.00385215  0.00247692  0.00171382] and c is 0.00040706083919732556 and cost is -1.962185494468386
The m is [-0.00444907 -0.00387502  0.0024907   0.00172503] and c is 0.00040926248873401525 and cost is -1.9619628222847925
The m is [-0.00447528 -0.00389792  0.00250455  0.00173629] and c is 0.00041150050842408964 and cost is -1.9617397593373678
The m is [-0.00450155 -0.00392081  0.00251847  0.00174764] and c is 0.000413674983490225 and cost is -1.961516178723167
The m is [-0.00452801 -0.00394384  0.0025326   0.00175917] and c is 0.000415759318144861 and cost is -1.9612907789499157
The m is [-0.00455447 -0.00396686  0.00254676  0.00177074] and c is 0.00041782395540679404 and cost is -1.9610652840671343
The m is [-0.00458115 -0.00399002  0.00256109  0.00178237] and c is 0.00042003213493741485 and cost is -1.9608380315460259
The m is [-0.00460789 

The m is [-0.00634872 -0.00552461  0.00350473  0.0025671 ] and c is 0.000584226851782028 and cost is -1.9458066027871683
The m is [-0.00638002 -0.00555168  0.00352125  0.00258123] and c is 0.0005871373507867692 and cost is -1.9455414396301507
The m is [-0.00641135 -0.00557885  0.00353781  0.00259534] and c is 0.0005899845097506506 and cost is -1.9452757734687744
The m is [-0.00644269 -0.00560601  0.00355436  0.00260945] and c is 0.0005928224651130375 and cost is -1.9450101571378715
The m is [-0.00647415 -0.00563333  0.00357105  0.00262346] and c is 0.0005955581866199609 and cost is -1.9447434131978993
The m is [-0.00650569 -0.00566073  0.00358779  0.00263757] and c is 0.0005983701937191758 and cost is -1.9444758779606177
The m is [-0.00653739 -0.00568827  0.00360462  0.0026519 ] and c is 0.0006010786130354864 and cost is -1.9442067449488634
The m is [-0.0065692  -0.00571593  0.0036215   0.00266638] and c is 0.0006036931699798305 and cost is -1.9439365038282883
The m is [-0.00660106 -0.

The m is [-0.00863581 -0.00751399  0.00474517  0.00360084] and c is 0.0007344422572849393 and cost is -1.9263772393465677
The m is [-0.00867214 -0.00754563  0.0047655   0.0036174 ] and c is 0.0007363762331475962 and cost is -1.9260679403884804
The m is [-0.00870847 -0.00757727  0.00478582  0.00363397] and c is 0.0007383259743470236 and cost is -1.9257586322911542
The m is [-0.00874489 -0.007609    0.00480628  0.00365054] and c is 0.0007403710653106262 and cost is -1.9254484018280924
The m is [-0.00878164 -0.00764091  0.004827    0.00366722] and c is 0.0007425622876770429 and cost is -1.9251357312767619
The m is [-0.00881846 -0.00767291  0.00484771  0.00368407] and c is 0.0007446735632782583 and cost is -1.9248222731125444
The m is [-0.00885532 -0.00770499  0.00486845  0.00370083] and c is 0.0007467047743669265 and cost is -1.9245084665560463
The m is [-0.00889219 -0.0077371   0.00488916  0.00371756] and c is 0.0007486966564352517 and cost is -1.9241946886904513
The m is [-0.0089291  -0

The m is [-0.01126807 -0.00981789  0.00623214  0.0047747 ] and c is 0.0008452093758565663 and cost is -1.904013169518183
The m is [-0.01130965 -0.00985439  0.00625557  0.00479264] and c is 0.0008476137642368118 and cost is -1.9036618173836521
The m is [-0.0113513  -0.00989097  0.00627901  0.00481064] and c is 0.0008499298940619181 and cost is -1.9033098465192764
The m is [-0.01139299 -0.00992758  0.00630249  0.00482861] and c is 0.0008521760526130942 and cost is -1.9029576172969114
The m is [-0.01143481 -0.00996428  0.00632601  0.0048467 ] and c is 0.000854350740556948 and cost is -1.902604405734497
The m is [-0.01147667 -0.010001    0.00634952  0.00486488] and c is 0.0008565720818600986 and cost is -1.9022508764533166
The m is [-0.01151856 -0.01003774  0.006373    0.00488309] and c is 0.0008587575621320034 and cost is -1.9018972313068565
The m is [-0.01156047 -0.01007452  0.00639647  0.00490131] and c is 0.0008608881380311043 and cost is -1.9015434147053367
The m is [-0.01160246 -0.01

The m is [-0.01420894 -0.01242062  0.00789215  0.00604636] and c is 0.0009663487230917572 and cost is -1.8791743960010496
The m is [-0.01425417 -0.01246092  0.00791796  0.00606582] and c is 0.0009673740252244975 and cost is -1.878792548192659
The m is [-0.01429955 -0.01250135  0.00794381  0.00608544] and c is 0.0009682915314494075 and cost is -1.878409454204545
The m is [-0.01434502 -0.01254183  0.00796969  0.00610514] and c is 0.0009691447773507081 and cost is -1.8780257177202941
The m is [-0.01439054 -0.01258237  0.0079956   0.00612495] and c is 0.0009699354201292788 and cost is -1.8776413847914855
The m is [-0.01443614 -0.012623    0.00802158  0.0061448 ] and c is 0.0009708059031990323 and cost is -1.8772563680303636
The m is [-0.01448198 -0.01266375  0.00804735  0.00616469] and c is 0.0009718115320595562 and cost is -1.8768704369521294
The m is [-0.014528   -0.01270466  0.00807322  0.00618468] and c is 0.0009727088882037955 and cost is -1.8764830374543504
The m is [-0.01457402 -0.0

The m is [-0.0174727  -0.01532209  0.00968919  0.00746255] and c is 0.0010705846747521511 and cost is -1.8518716426170212
The m is [-0.01752301 -0.01536673  0.00971686  0.00748431] and c is 0.0010725817404908702 and cost is -1.8514530605357202
The m is [-0.01757341 -0.01541142  0.0097446   0.00750618] and c is 0.001074655484883422 and cost is -1.851033736037368
The m is [-0.01762399 -0.01545625  0.00977234  0.00752821] and c is 0.0010768548621281492 and cost is -1.8506131305391877
The m is [-0.01767458 -0.01550113  0.00980007  0.00755021] and c is 0.001079005904888587 and cost is -1.8501924346181788
The m is [-0.01772526 -0.01554612  0.00982789  0.00757213] and c is 0.001081238241173758 and cost is -1.8497710627237667
The m is [-0.01777615 -0.01559123  0.0098556   0.00759414] and c is 0.0010836056188652289 and cost is -1.8493486017794458
The m is [-0.01782714 -0.01563646  0.00988342  0.00761621] and c is 0.0010860518277813085 and cost is -1.8489251838414753
The m is [-0.0178782  -0.015

The m is [-0.02105606 -0.01850087  0.0116386   0.00901775] and c is 0.0012217035969197562 and cost is -1.822234520946399
The m is [-0.02111126 -0.01854968  0.01166847  0.00904214] and c is 0.0012235590920375826 and cost is -1.821780354406871
The m is [-0.02116647 -0.01859853  0.01169835  0.00906651] and c is 0.0012254497538983081 and cost is -1.821326114733445
The m is [-0.02122174 -0.0186475   0.01172834  0.00909077] and c is 0.001227250884774324 and cost is -1.8208713275312882
The m is [-0.02127701 -0.01869648  0.01175835  0.00911502] and c is 0.001229068078505634 and cost is -1.8204164346147764
The m is [-0.02133239 -0.01874561  0.01178858  0.00913919] and c is 0.001230766674110866 and cost is -1.819960483962454
The m is [-0.02138782 -0.01879484  0.01181892  0.00916338] and c is 0.001232406390696978 and cost is -1.8195038664166492
The m is [-0.02144326 -0.01884408  0.01184926  0.00918758] and c is 0.0012340234651377385 and cost is -1.8190472204840369
The m is [-0.0214987  -0.0188933

The m is [-0.02495777 -0.02196349  0.01376178  0.01071834] and c is 0.0013509043680284035 and cost is -1.7902688563595421
The m is [-0.02501782 -0.0220168   0.01379425  0.01074401] and c is 0.0013530898065412294 and cost is -1.7897805501400108
The m is [-0.02507806 -0.02207031  0.01382682  0.0107697 ] and c is 0.0013554022133549775 and cost is -1.789290789034883
The m is [-0.02513845 -0.02212396  0.01385946  0.01079556] and c is 0.0013576062220998005 and cost is -1.788799754913083
The m is [-0.02519901 -0.02217775  0.01389214  0.01082156] and c is 0.0013596886506859925 and cost is -1.788307398241691
The m is [-0.02525958 -0.02223154  0.01392482  0.01084753] and c is 0.0013617549257095454 and cost is -1.7878151958862483
The m is [-0.02532033 -0.02228547  0.01395753  0.01087348] and c is 0.0013639534759387478 and cost is -1.787321935292685
The m is [-0.02538137 -0.02233959  0.01399021  0.01089965] and c is 0.0013663176534687966 and cost is -1.7868268115351098
The m is [-0.02544251 -0.022

The m is [-0.03030299 -0.02671724  0.01665481  0.01294782] and c is 0.001556256574598914 and cost is -1.7471598154819186
The m is [-0.03036967 -0.02677657  0.01669092  0.01297561] and c is 0.0015585251188983225 and cost is -1.7466258744015482
The m is [-0.03043649 -0.02683606  0.01672712  0.01300349] and c is 0.0015607101823540363 and cost is -1.7460907437330992
The m is [-0.03050336 -0.0268956   0.01676335  0.01303139] and c is 0.001562953457783715 and cost is -1.745555252236254
The m is [-0.03057029 -0.02695523  0.01679962  0.01305929] and c is 0.001565265121402549 and cost is -1.7450193055684275
The m is [-0.03063725 -0.02701488  0.01683591  0.01308721] and c is 0.0015676169101106583 and cost is -1.7444832119508864
The m is [-0.03070421 -0.02707455  0.01687219  0.01311512] and c is 0.001569974503564952 and cost is -1.7439471918568328
The m is [-0.03077136 -0.02713433  0.01690865  0.01314315] and c is 0.001572449592871317 and cost is -1.743409730184644
The m is [-0.03083851 -0.027194

The m is [-0.03498142 -0.0308807   0.01916788  0.01490515] and c is 0.0017431235097831047 and cost is -1.7099584815134332
The m is [-0.03505254 -0.03094395  0.01920544  0.01493522] and c is 0.00174613465415407 and cost is -1.709397534640955
The m is [-0.03512378 -0.03100731  0.01924313  0.0149652 ] and c is 0.0017492523028712096 and cost is -1.7088358204343632
The m is [-0.03519517 -0.03107076  0.01928088  0.01499533] and c is 0.0017524736545365628 and cost is -1.7082730608086547
The m is [-0.03526667 -0.03113429  0.01931873  0.01502556] and c is 0.0017557739895453636 and cost is -1.707709460150183
The m is [-0.0353382  -0.03119788  0.01935661  0.0150558 ] and c is 0.0017591179646973738 and cost is -1.7071456382181835
The m is [-0.03540982 -0.03126155  0.0193945   0.01508616] and c is 0.0017623843621270288 and cost is -1.7065811687911658
The m is [-0.03548145 -0.03132523  0.01943241  0.01511653] and c is 0.0017656689207568525 and cost is -1.7060166701252277
The m is [-0.03555323 -0.031

The m is [-0.04132373 -0.03650656  0.02254009  0.01755435] and c is 0.0019825151450406087 and cost is -1.6604143499957849
The m is [-0.04139983 -0.03657399  0.02258078  0.01758581] and c is 0.001985307901303687 and cost is -1.6598254275235829
The m is [-0.04147597 -0.03664148  0.0226215   0.01761727] and c is 0.001988151178687594 and cost is -1.6592362676563515
The m is [-0.04155219 -0.03670908  0.02266218  0.01764884] and c is 0.0019909197791537366 and cost is -1.658646481446999
The m is [-0.04162841 -0.03677669  0.02270287  0.01768042] and c is 0.0019936880208685213 and cost is -1.658056821337394
The m is [-0.04170466 -0.0368443   0.02274357  0.01771202] and c is 0.0019964142745394283 and cost is -1.6574670820723727
The m is [-0.04178092 -0.03691192  0.02278428  0.01774362] and c is 0.001999142861537975 and cost is -1.6568774646063191
The m is [-0.04185734 -0.0369797   0.02282512  0.01777527] and c is 0.002001983266561439 and cost is -1.6562865681004646
The m is [-0.04193383 -0.03704

The m is [-0.04666105 -0.04124456  0.02537093  0.01975484] and c is 0.0021555949486239748 and cost is -1.619435724483953
The m is [-0.04674197 -0.04131673  0.02541393  0.01978771] and c is 0.002157809515987329 and cost is -1.6188186067384749
The m is [-0.04682291 -0.04138894  0.02545691  0.01982053] and c is 0.0021599717032150124 and cost is -1.618201589058314
The m is [-0.04690397 -0.04146125  0.02550004  0.0198535 ] and c is 0.0021622156903344353 and cost is -1.6175835542504848
The m is [-0.04698506 -0.04153362  0.02554308  0.01988649] and c is 0.002164515645396141 and cost is -1.6169654622943133
The m is [-0.04706622 -0.04160602  0.02558615  0.01991957] and c is 0.002166762431638858 and cost is -1.616346964820245
The m is [-0.04714743 -0.04167851  0.02562926  0.01995261] and c is 0.0021690681638134967 and cost is -1.615728220972042
The m is [-0.04722865 -0.04175102  0.02567236  0.01998563] and c is 0.0021713957240318167 and cost is -1.615109570374681
The m is [-0.04730991 -0.0418236

The m is [-0.05228882 -0.0462855   0.02834544  0.02204432] and c is 0.0023078556846099327 and cost is -1.5768143386145737
The m is [-0.05237346 -0.04636157  0.02838989  0.02207863] and c is 0.0023098981553913863 and cost is -1.57617845161681
The m is [-0.05245813 -0.04643767  0.02843435  0.02211299] and c is 0.0023119776697661768 and cost is -1.5755424708451589
The m is [-0.05254285 -0.04651385  0.02847888  0.02214735] and c is 0.0023139856132722283 and cost is -1.574906051233588
The m is [-0.05262757 -0.04659003  0.02852344  0.02218167] and c is 0.0023159741098164673 and cost is -1.574269820715841
The m is [-0.05271247 -0.04666632  0.02856805  0.02221608] and c is 0.00231808560476995 and cost is -1.5736326267694285
The m is [-0.05279746 -0.04674277  0.02861274  0.02225036] and c is 0.002320101265588193 and cost is -1.5729948098860607
The m is [-0.05288251 -0.04681924  0.02865756  0.02228464] and c is 0.002322177622548332 and cost is -1.572356675695022
The m is [-0.05296771 -0.04689584

The m is [-0.05821163 -0.05161465  0.03150386  0.02443851] and c is 0.0024681515072368537 and cost is -1.53259836186278
The m is [-0.05830091 -0.05169502  0.03155174  0.02447483] and c is 0.002470826052427286 and cost is -1.531936801148502
The m is [-0.05839026 -0.05177545  0.03159977  0.02451119] and c is 0.0024735805399026837 and cost is -1.5312747410912746
The m is [-0.0584796  -0.05185594  0.0316478   0.02454752] and c is 0.0024762796215004766 and cost is -1.5306127095441195
The m is [-0.05856909 -0.05193658  0.03169591  0.02458397] and c is 0.002478882051430189 and cost is -1.529949640335047
The m is [-0.05865859 -0.05201724  0.03174402  0.02462042] and c is 0.0024814534703227007 and cost is -1.529286641778691
The m is [-0.05874815 -0.05209801  0.03179216  0.02465683] and c is 0.002483948115545236 and cost is -1.5286232806316462
The m is [-0.05883779 -0.05217893  0.03184044  0.02469316] and c is 0.0024863525943660527 and cost is -1.5279592442705352
The m is [-0.05892744 -0.0522598

The m is [-0.06442294 -0.05720672  0.03483892  0.02694876] and c is 0.0026574118515891787 and cost is -1.4870005032240037
The m is [-0.0645162  -0.05729062  0.03488912  0.02698662] and c is 0.0026602975132644004 and cost is -1.4863220226772187
The m is [-0.06460947 -0.05737455  0.03493935  0.02702446] and c is 0.0026631529017382203 and cost is -1.4856436106005115
The m is [-0.06470277 -0.05745849  0.03498957  0.02706229] and c is 0.0026659756160545382 and cost is -1.4849652792997157
The m is [-0.06479609 -0.05754247  0.0350398   0.02710014] and c is 0.002668842921116655 and cost is -1.4842869222779214
The m is [-0.06488952 -0.05762655  0.03509006  0.02713804] and c is 0.0026716350056785374 and cost is -1.4836079998115121
The m is [-0.06498304 -0.0577107   0.03514054  0.02717598] and c is 0.002674337362305219 and cost is -1.4829283594490146
The m is [-0.06507673 -0.05779499  0.0351911   0.02721403] and c is 0.0026769370849339213 and cost is -1.4822476307227601
The m is [-0.06517047 -0.0

The m is [-0.0708932  -0.06304607  0.0383431   0.02953957] and c is 0.0028260559288749987 and cost is -1.4403337813045864
The m is [-0.07099011 -0.06313377  0.03839611  0.02957819] and c is 0.0028281189451006133 and cost is -1.4396404318691582
The m is [-0.0718661  -0.06392677  0.03887535  0.02992671] and c is 0.002845468437449335 and cost is -1.4333821667752862
The m is [-0.07196373 -0.06401519  0.03892877  0.02996549] and c is 0.0028472007554151816 and cost is -1.4326857028596045
The m is [-0.07206144 -0.06410369  0.03898212  0.0300043 ] and c is 0.002848847307548767 and cost is -1.431989013313725
The m is [-0.07215914 -0.0641922   0.03903545  0.03004306] and c is 0.0028504674922730897 and cost is -1.4312926190376556
The m is [-0.07225692 -0.06428076  0.03908887  0.0300819 ] and c is 0.002852156791443419 and cost is -1.4305957960202849
The m is [-0.07235471 -0.06436934  0.03914225  0.03012078] and c is 0.0028538229635477054 and cost is -1.4298990927497695
The m is [-0.07245252 -0.064

The m is [-0.07842661 -0.06986335  0.04244839  0.03253729] and c is 0.0029504234778747044 and cost is -1.3870796978875286
The m is [-0.07852828 -0.06995523  0.04250351  0.03257774] and c is 0.0029521647942313254 and cost is -1.3863703135420686
The m is [-0.07862999 -0.07004715  0.04255869  0.03261816] and c is 0.0029538371518509385 and cost is -1.3856608410426816
The m is [-0.07873175 -0.07013909  0.04261384  0.0326586 ] and c is 0.0029554478799537287 and cost is -1.3849514241174692
The m is [-0.07883354 -0.07023106  0.04266899  0.03269909] and c is 0.002957087454067091 and cost is -1.3842419814443574
The m is [-0.07893534 -0.07032305  0.04272416  0.03273954] and c is 0.0029587455980981457 and cost is -1.383532737768313
The m is [-0.07903725 -0.07041511  0.04277938  0.03277997] and c is 0.0029604830916063556 and cost is -1.3828230707114022
The m is [-0.07913934 -0.0705073   0.04283483  0.03282058] and c is 0.002962115805181542 and cost is -1.38211217110432
The m is [-0.07924147 -0.0705

The m is [-0.08546591 -0.07623523  0.04625963  0.03531581] and c is 0.003085355989643024 and cost is -1.3385020975357946
The m is [-0.08557109 -0.07633074  0.04631654  0.03535682] and c is 0.003087679384302959 and cost is -1.3377842883775244
The m is [-0.08567632 -0.07642628  0.04637349  0.03539788] and c is 0.0030900558683070174 and cost is -1.3370663045889484
The m is [-0.08578171 -0.07652193  0.04643037  0.03543911] and c is 0.0030925332494250164 and cost is -1.3363477390145495
The m is [-0.08588715 -0.07661764  0.04648728  0.03548036] and c is 0.003094942455191518 and cost is -1.3356290171380496
The m is [-0.08599262 -0.07671338  0.04654423  0.03552157] and c is 0.00309739179650369 and cost is -1.33491034712569
The m is [-0.08609839 -0.0768093   0.04660126  0.03556287] and c is 0.0030999810090695874 and cost is -1.3341903104872408
The m is [-0.08620426 -0.07690531  0.04665833  0.03560417] and c is 0.0031024784028951895 and cost is -1.333469879254425
The m is [-0.08631016 -0.0770013

The m is [-0.09277011 -0.0828684   0.05019294  0.03817313] and c is 0.0032173654089644177 and cost is -1.289242975490598
The m is [-0.09287975 -0.082968    0.05025214  0.03821608] and c is 0.0032191567926298497 and cost is -1.2885120529931902
The m is [-0.09298948 -0.08306772  0.05031148  0.03825896] and c is 0.003220867329372353 and cost is -1.2877806368022318
The m is [-0.09309929 -0.08316753  0.05037081  0.03830197] and c is 0.0032225083029944917 and cost is -1.2870488621046081
The m is [-0.09320924 -0.08326747  0.05043035  0.03834499] and c is 0.0032242379990073565 and cost is -1.2863163086937222
The m is [-0.09331925 -0.08336746  0.05049001  0.03838802] and c is 0.0032260481069874107 and cost is -1.2855834601189178
The m is [-0.09342933 -0.08346755  0.05054972  0.0384311 ] and c is 0.0032278105636858903 and cost is -1.2848503186920515
The m is [-0.0935394  -0.08356764  0.05060943  0.03847418] and c is 0.0032295760221733947 and cost is -1.2841174357061078
The m is [-0.0936495  -0.0

The m is [-0.1003584  -0.0897698   0.0542859   0.04113455] and c is 0.0033324663430714654 and cost is -1.2392707442288122
The m is [-0.10047173 -0.08987302  0.05434682  0.04117876] and c is 0.0033341085734915187 and cost is -1.2385338698458301
The m is [-0.10058536 -0.08997641  0.05440772  0.04122302] and c is 0.00333590211579479 and cost is -1.2377959539440466
The m is [-0.10069907 -0.09007985  0.05446864  0.04126733] and c is 0.0033376471735024754 and cost is -1.2370578804189698
The m is [-0.10081278 -0.0901833   0.05452957  0.04131165] and c is 0.0033394001510937393 and cost is -1.2363200200849103
The m is [-0.10092657 -0.09028682  0.05459057  0.04135595] and c is 0.003341085030451957 and cost is -1.2355819348163048
The m is [-0.10104056 -0.0903905   0.05465168  0.04140021] and c is 0.0033428871338332515 and cost is -1.2348430725202835
The m is [-0.10115454 -0.09049415  0.0547128   0.04144448] and c is 0.0033447115447286157 and cost is -1.234104529871511
The m is [-0.10126853 -0.090

The m is [-0.10821634 -0.09692516  0.05849499  0.04419395] and c is 0.0034481746959587976 and cost is -1.1888752322832052
The m is [-0.10833395 -0.09703239  0.058558    0.04423985] and c is 0.0034495787612961845 and cost is -1.188130725298547
The m is [-0.10845155 -0.09713961  0.058621    0.04428575] and c is 0.0034509893671477027 and cost is -1.1873865319335732
The m is [-0.1085692  -0.09724688  0.058684    0.04433173] and c is 0.0034523504873938985 and cost is -1.18664236327947
The m is [-0.10868695 -0.09735426  0.05874704  0.04437775] and c is 0.003453643760571547 and cost is -1.1858978033164294
The m is [-0.10880472 -0.09746168  0.05881012  0.0444238 ] and c is 0.003454902930053148 and cost is -1.1851532836456296
The m is [-0.10892249 -0.09756909  0.0588732   0.04446985] and c is 0.0034561594892010003 and cost is -1.1844090701102181
The m is [-0.1090403  -0.09767654  0.05893632  0.0445159 ] and c is 0.0034573584636711334 and cost is -1.1836649154716274
The m is [-0.10915815 -0.0977

The m is [-0.11630595 -0.10429206  0.06283373  0.04736296] and c is 0.0035204260248243653 and cost is -1.1383716105547939
The m is [-0.11642662 -0.10440168  0.06289874  0.04741067] and c is 0.003521524113424149 and cost is -1.137628906967358
The m is [-0.11654729 -0.10451133  0.06296378  0.0474584 ] and c is 0.0035226387162602207 and cost is -1.1368863998898497
The m is [-0.116668   -0.10462097  0.06302884  0.04750613] and c is 0.0035237924508892743 and cost is -1.1361441128582066
The m is [-0.11678872 -0.10473062  0.06309391  0.04755383] and c is 0.003524979858230697 and cost is -1.1354021103912755
The m is [-0.11690952 -0.10484033  0.063159    0.04760162] and c is 0.00352611724468772 and cost is -1.1346599067572125
The m is [-0.11703039 -0.10495003  0.06322417  0.0476495 ] and c is 0.003527208628988331 and cost is -1.1339176952633236
The m is [-0.11715125 -0.10505971  0.06328936  0.04769737] and c is 0.003528296411388236 and cost is -1.133175807074706
The m is [-0.11727212 -0.1051694

The m is [-0.12463055 -0.11183798  0.06731739  0.05065379] and c is 0.00360105621500796 and cost is -1.0879248195099946
The m is [-0.1247552  -0.11195066  0.06738449  0.05070297] and c is 0.003602611106188896 and cost is -1.0871818554090276
The m is [-0.1248799  -0.11206338  0.06745167  0.05075216] and c is 0.003604205808905195 and cost is -1.0864388624087942
The m is [-0.12500459 -0.11217612  0.06751888  0.05080133] and c is 0.003605788128500419 and cost is -1.0856961690963611
The m is [-0.12512936 -0.11228891  0.06758614  0.05085059] and c is 0.0036074391815380928 and cost is -1.0849533203671262
The m is [-0.12525419 -0.11240178  0.06765332  0.05089991] and c is 0.0036091556772066087 and cost is -1.0842105242559856
The m is [-0.12537912 -0.11251473  0.06772057  0.05094932] and c is 0.0036109577737156984 and cost is -1.08346740696375
The m is [-0.12550415 -0.11262778  0.06778795  0.05099875] and c is 0.003612692213081087 and cost is -1.0827239388902303
The m is [-0.12562918 -0.1127408

The m is [-0.13335673 -0.1197258   0.07203459  0.05410069] and c is 0.003701260059765486 and cost is -1.0367008303327439
The m is [-0.13348503 -0.1198416   0.07210418  0.05415128] and c is 0.0037025225140094086 and cost is -1.0359602777410524
The m is [-0.13361343 -0.11995747  0.07217393  0.05420194] and c is 0.003703859653284068 and cost is -1.0352193635932752
The m is [-0.13374183 -0.12007335  0.0722437   0.05425261] and c is 0.003705210798406463 and cost is -1.0344787342857156
The m is [-0.13387024 -0.12018922  0.07231348  0.05430328] and c is 0.003706554716832903 and cost is -1.0337384804892455
The m is [-0.1339987  -0.12030515  0.07238335  0.05435394] and c is 0.0037079422521356016 and cost is -1.0329982064232002
The m is [-0.13412737 -0.12042121  0.07245336  0.05440463] and c is 0.0037094461822277415 and cost is -1.0322572376909003
The m is [-0.13425604 -0.12053728  0.07252335  0.05445532] and c is 0.0037109370786277116 and cost is -1.0315166337456008
The m is [-0.13438472 -0.120

The m is [-0.14218557 -0.12770335  0.07684277  0.05756715] and c is 0.0037853047838867205 and cost is -0.986544226376247
The m is [-0.14231693 -0.12782222  0.07691431  0.05761856] and c is 0.003786019729257025 and cost is -0.9858105549286879
The m is [-0.1424483  -0.1279411   0.07698584  0.05766999] and c is 0.0037867132245045414 and cost is -0.9850771916089576
The m is [-0.14257978 -0.12806009  0.07705747  0.05772141] and c is 0.0037873232626869525 and cost is -0.9843435362637889
The m is [-0.14271129 -0.12817911  0.07712914  0.05777277] and c is 0.0037878754132094055 and cost is -0.9836101366393616
The m is [-0.14284284 -0.12829818  0.07720082  0.0578242 ] and c is 0.003788380561445406 and cost is -0.9828768459171476
The m is [-0.14297445 -0.12841735  0.07727256  0.05787562] and c is 0.003788820518115724 and cost is -0.9821434662846633
The m is [-0.14310606 -0.12853651  0.07734431  0.05792705] and c is 0.0037892684990358952 and cost is -0.9814104704289378
The m is [-0.1432378  -0.128

The m is [-0.15121957 -0.13589139  0.08176299  0.06110418] and c is 0.003817037092632524 and cost is -0.9369341470315179
The m is [-0.15135433 -0.13601341  0.08183596  0.06115733] and c is 0.003817655500149065 and cost is -0.9362081085463102
The m is [-0.15148918 -0.1361355   0.08190897  0.06121055] and c is 0.0038183555328242915 and cost is -0.9354819426619261
The m is [-0.15162404 -0.1362576   0.08198198  0.06126378] and c is 0.0038190485034384244 and cost is -0.9347561765515721
The m is [-0.15175894 -0.13637973  0.08205494  0.06131709] and c is 0.003819795234189467 and cost is -0.9340305877206686
The m is [-0.15189392 -0.13650192  0.08212799  0.06137049] and c is 0.003820589487002822 and cost is -0.9333049221459836
The m is [-0.15202898 -0.13662416  0.08220101  0.06142394] and c is 0.0038214460560819135 and cost is -0.9325793411438852
The m is [-0.15216427 -0.13674655  0.08227402  0.06147756] and c is 0.0038224280179333333 and cost is -0.9318531442798123
The m is [-0.15229958 -0.136

The m is [-0.16049702 -0.14429515  0.08678778  0.06474531] and c is 0.003866445216906105 and cost is -0.8878800150378703
The m is [-0.16063509 -0.14442044  0.08686243  0.06479894] and c is 0.003867096671705286 and cost is -0.88716427212236
The m is [-0.16077315 -0.14454575  0.08693704  0.06485254] and c is 0.0038677702239401828 and cost is -0.8864489532463464
The m is [-0.16091127 -0.14467111  0.08701167  0.06490623] and c is 0.003868491387696408 and cost is -0.8857337218605948
The m is [-0.16104947 -0.14479654  0.0870865   0.06495998] and c is 0.003869130247431567 and cost is -0.8850183204327311
The m is [-0.16118769 -0.14492196  0.08716137  0.06501368] and c is 0.003869726943121922 and cost is -0.8843033058013862
The m is [-0.16132594 -0.14504741  0.08723626  0.06506739] and c is 0.003870293621901658 and cost is -0.8835885236536432
The m is [-0.16146421 -0.1451729   0.08731113  0.06512114] and c is 0.0038708926453461974 and cost is -0.8828740292376033
The m is [-0.16160252 -0.1452984

The m is [-0.16998836 -0.15291167  0.09192876  0.06844671] and c is 0.00389844477097438 and cost is -0.8396156500457769
The m is [-0.17012954 -0.15303976  0.09200514  0.06850187] and c is 0.0038990232331332334 and cost is -0.8389127580342413
The m is [-0.17027073 -0.15316785  0.09208154  0.068557  ] and c is 0.0038995714978332895 and cost is -0.8382102957655504
The m is [-0.17041193 -0.15329596  0.09215793  0.06861215] and c is 0.003900093787295597 and cost is -0.8375081720916533
The m is [-0.1705532  -0.15342412  0.09223436  0.06866737] and c is 0.0039005804118809966 and cost is -0.8368061636391394
The m is [-0.17069454 -0.15355236  0.09231086  0.06872252] and c is 0.003900996500562763 and cost is -0.8361042356240806
The m is [-0.1708359  -0.15368069  0.09238738  0.06877766] and c is 0.0039013497461869483 and cost is -0.8354024917673675
The m is [-0.17097732 -0.15380904  0.09246394  0.06883286] and c is 0.0039016746301376148 and cost is -0.8347009239568285
The m is [-0.17111873 -0.153

The m is [-0.1796852  -0.16170469  0.09716252  0.07223913] and c is 0.003910376413365792 and cost is -0.7923688699809373
The m is [-0.17982961 -0.16183534  0.09724048  0.07229549] and c is 0.003910777183139105 and cost is -0.7916813846030084
The m is [-0.17997407 -0.16196606  0.09731849  0.07235185] and c is 0.003911135409718624 and cost is -0.7909940738282161
The m is [-0.18011856 -0.16209681  0.09739652  0.07240816] and c is 0.003911526704854706 and cost is -0.7903070988885795
The m is [-0.18026304 -0.16222756  0.09747456  0.07246448] and c is 0.003911925109985237 and cost is -0.7896205823741674
The m is [-0.18040751 -0.1623583   0.09755262  0.07252079] and c is 0.003912301850134039 and cost is -0.7889345485442705
The m is [-0.18055202 -0.16248908  0.09763069  0.07257715] and c is 0.003912717926146317 and cost is -0.7882487731247501
The m is [-0.18069655 -0.16261988  0.09770877  0.07263353] and c is 0.003913113007911943 and cost is -0.7875633401246277
The m is [-0.18084135 -0.1627508

The m is [-0.18962415 -0.17069061  0.10249318  0.07612042] and c is 0.003954157507801769 and cost is -0.746157008570354
The m is [-0.18977213 -0.17082443  0.10257241  0.07617798] and c is 0.003955054511390703 and cost is -0.7454853899136845
The m is [-0.18992014 -0.17095829  0.1026517   0.07623555] and c is 0.003955986344875755 and cost is -0.7448140394402575
The m is [-0.19006818 -0.1710922   0.10273102  0.07629312] and c is 0.0039568906209508274 and cost is -0.7441429842543897
The m is [-0.19021621 -0.1712261   0.10281033  0.0763507 ] and c is 0.003957788267745544 and cost is -0.7434724333310381
The m is [-0.19036425 -0.17136003  0.10288964  0.0764083 ] and c is 0.003958660201650468 and cost is -0.7428022756214859
The m is [-0.1905123  -0.17149395  0.10296895  0.0764659 ] and c is 0.003959533240390078 and cost is -0.7421325999126955
The m is [-0.19066038 -0.17162791  0.10304828  0.0765235 ] and c is 0.003960440083855496 and cost is -0.741463231144088
The m is [-0.19080848 -0.17176188

### optimal values i got through stochastic Gradient Descent is The m is [-0.40710427 -0.36256501  0.21303677  0.15584904  0.01167379] and cost is -0.26819551574655887 after this value my algorithm started increasing its cost.Basically thats the nature of Gradient Descent.


### Applying  mini Batch Gradient Descent to see the change of cost function with a little variation in my algorithm...I choose my batch size to be 100.

In [36]:
def Mini_Batch_cost(data,m,c):
    cost=0
    for i in range(len(data)):
        x=np.array(data.iloc[i][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']])
        y=np.array(data.iloc[i]['Electrical_output'])
        cost=cost+(-2/len(data))*(y-np.sum(m*x)-c)**2
    return cost

In [43]:
def step_MiniBatch_gradient(data,learning_rate,m,c):
    M=len(data)
    m_slope=np.zeros(4)
    c_slope=0
    Batch_size=100
    count=0
    prevcost=1000
    for i in range(len(data)):
        x=np.array(data.iloc[i][['Temperature','Ambient_pressure','Relative_humidity','Exhaust_vaccume']])
        y=np.array(data.iloc[i]['Electrical_output'])
        m_slope=m_slope+((-2/M)*(y-np.sum(m*x))*x)
        c_slope=c_slope+((-2/M)*(y-np.sum(m*x)))
        count=count+1
        if count==Batch_size:
            m=m-(learning_rate*m_slope)
            c=c-(learning_rate*c_slope)
            count=0
            print("The m is {} and c is {} and cost is {}".format(m,c,Mini_Batch_cost(data,m,c)))
       
       
   
         
    return m,c

In [46]:
def Mini_Batch_gd(data):
    no_of_iterations=100
    learning_rate=0.01
    m=np.zeros(4)
    c=0
    m,c=step_MiniBatch_gradient(data,learning_rate,m,c)
    return m,c

In [47]:
new_m,new_c=Mini_Batch_gd(data)
print("The new_m is {} and new_c is{}".format(new_m,new_c))
new_m,new_c

The m is [-0.00032442 -0.00028285  0.00019225  0.0001034 ] and c is 3.710814506111633e-05 and cost is -1.997224090729324
The m is [-0.00091189 -0.00079053  0.00051301  0.00031686] and c is 8.185382605306237e-05 and cost is -1.9922344160792358
The m is [-0.00175686 -0.00153089  0.00097114  0.00067494] and c is 0.00014217071317581084 and cost is -1.9849619989923104
The m is [-0.00288703 -0.002513    0.0015716   0.00118116] and c is 0.00024286702045185214 and cost is -1.9752701824424006
The m is [-0.00427938 -0.00373351  0.00234891  0.00180388] and c is 0.00029511332475709416 and cost is -1.9632559477068994
The m is [-0.00590626 -0.00518375  0.00327882  0.00250554] and c is 0.00033114641071432766 and cost is -1.9491851456876144
The m is [-0.00780431 -0.00686678  0.00432097  0.00333226] and c is 0.00041326622674314024 and cost is -1.9329489062127405
The m is [-0.00995503 -0.00877897  0.00548587  0.00426979] and c is 0.00048068332600836216 and cost is -1.9146609655407472
The m is [-0.012399

The m is [-0.49281126 -0.44679495  0.25847177  0.18199325] and c is 0.0030765420133414883 and cost is -0.35156204164796817
The m is [-0.50323712 -0.45618878  0.26338854  0.18514167] and c is 0.0030699768693174434 and cost is -0.37166272876725465
The m is [-0.51362987 -0.46554965  0.26822909  0.18823858] and c is 0.003064874249896265 and cost is -0.39377106524046174
The new_m is [-0.51362987 -0.46554965  0.26822909  0.18823858] and new_c is0.003064874249896265


(array([-0.51362987, -0.46554965,  0.26822909,  0.18823858]),
 0.0030648742498962651)

### We are heading towars the optima very fastly than batch gradient descent but it is not global optima from the learning rate of 0.01 i got to see The m is [-0.46133466 -0.41838976  0.24335982  0.17221593] and c is 0.0029938644643197334 and cost is -0.3037181966971105 as my optimal values from the above result


In [49]:
print("The Mini batch Gradient_values are m {} and c {}".format(new_m,new_c))

The Mini batch Gradient_values are m [-0.51362987 -0.46554965  0.26822909  0.18823858] and c 0.003064874249896265


## Observations:
### 1.Gradient Descent is sensitive to lower values.
### 2.Here i applied column standardiastion in feature scaling over column normalisation because i feel column standardisation follows the Normal Distribution N(0,1) and almost majority of the real world data follows normal distribution.
### 3.Decreasing learning_rate will give optimal coefficients but we have to choose the number of iterations should also be more.
### 4.I did not try to change the learning_rate in the mid of the algorithm but i read Adaptive learning rate will give better results.I will definitely try it in my future projects.
### 5.I always tried to compare my optimal values with sklearn linear_model.LinearRegression() to see how my values over sklearn and wanted to verify the correctness.I see for Batch Gradient Descent with column standardisation my values are equal with sklearn .So i see even sklearn used Batch Gradient Descent to solve my problem